modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngineService.java

```diff
@@ -109,6 +110,11 @@ public String getExtension() {
 
     @Override
     public Object compile(String scriptName, final String scriptSource, final Map<String, String> params) {
+        return compile(GenericElasticsearchScript.class, scriptName, scriptSource, params, GenericElasticsearchScript.DERIVED_ARGUMENTS);
+    }
+
+    <T> T compile(Class<T> iface, String scriptName, final String scriptSource, final Map<String, String> params,
```

Comment:
Package private because I want to test with this. At some point I think we should yank some of this code into another spot that isn't Elasticsearch specific.